src/RoslynPad.Common.UI/ViewModels/DocumentViewModel.cs

```diff
@@ -160,17 +185,80 @@ private ObservableCollection<DocumentViewModel> ReadChildren()
         {
             return new ObservableCollection<DocumentViewModel>(
                 IOUtilities.EnumerateDirectories(Path)
-                .Select(x => new DocumentViewModel(x, isFolder: true))
-                .OrderBy(OrderByName)
-                    .Concat(IOUtilities.EnumerateFiles(Path, "*" + DefaultFileExtension)
-                        .Select(x => new DocumentViewModel(x, isFolder: false))
-                        .Where(x => !x.IsAutoSave)
-                        .OrderBy(OrderByName)));
+                    .Concat(IOUtilities.EnumerateFiles(Path, "*" + DefaultFileExtension))
+                    .Select(x => new DocumentViewModel(x, _documentFileWatcher))
+                    .Where(x => !x.IsAutoSave)
+                    .OrderBy(dd => !dd.IsFolder)
+                    .ThenBy(OrderByName));
         }
 
         private static string OrderByName(DocumentViewModel x)
         {
             return Regex.Replace(x.Name, "[0-9]+", m => m.Value.PadLeft(100, '0'));
         }
+
+        private static Func<IEnumerable<DocumentViewModel>, IOrderedEnumerable<DocumentViewModel>> SortPredicate =>
+            d => d.OrderBy (dd => !dd.IsFolder).ThenBy (OrderByName);
+
+        public void OnDocumentFileChanged (DocumentFileChanged value)
+        {
+            if (!IsFolder && value.Path != Path)
+            {
+                return;
+            }
+            if (IsFolder && value.Path != Path && System.IO.Path.GetDirectoryName(value.Path) != Path)
+            {
+                return;
+            }
+
+            switch (value.Type)
+            {
+            case DocumentFileChangeType.Created:
+                OnDocumentCreated(value);
+                break;
+            case DocumentFileChangeType.Deleted:
+                OnDocumentDeleted(value);
+                break;
+            case DocumentFileChangeType.Renamed:
+                OnDocumentRenamed(value);
+                break;
+            }
+        }
+
+        private void OnDocumentRenamed (DocumentFileChanged value)
+        {
+            if (Path != value.Path)
+                return; //Rename only applies to self
+            Path = value.NewPath;
+            Name = System.IO.Path.GetFileName(value.NewPath);
+        }
+
+        private void OnDocumentCreated (DocumentFileChanged value)
+        {
+            if (!IsFolder)
```

Comment:
Also check `IsChildrenInitialized` - again we don't want them to be initialized by accident.